numpy/testing/utils.py

```diff
@@ -1937,3 +1964,150 @@ def __exit__(self, *exc_info):
                 mod.__warningregistry__.clear()
             if mod in self._warnreg_copies:
                 mod.__warningregistry__.update(self._warnreg_copies[mod])
+
+
+class suppressed_warnings(object):
+    """
+    Context manager doing much the same as ``warnings.catch_warnings``.
+    However, it also provides a suppression mechanism, meaning that instead
+    of setting a warning to "ignore", it is possible to add a suppression.
+    This has the advantage of ensuring a clean warning registry.
+
+    Suppressions are inserted either immidiatly, or upon entering the
+    context in the order they were added, so that they are applied in
+    reversed order. A suppression is equivalent to an "always" but filters
+    the warning instead of printing it.
+
+    Note that all of this is only necessary for Python versions before
+    python (about) 3.4. After that "ingnore" can be used safely within
+    the catch_warnings context.
```

Comment:
This is all about the `__warningregistry__` and avoiding getting stuff into it, which catch_warnings does not really help with at all. On newer pythons, this is solved (the registry is invalidated automatically).